### ISSUE TITLE
Get rid of "parser error : Premature end of data in tag ..." error (latexmk involved)

### ISSUE THREAD

**Issue by dbitouze** [plk/biblatex#348]
Consider the following (wrong, see `\textbf{` not closed) `test.tex` file:

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage[backend=biber]{biblatex}%
%
\addbibresource{biblatex-examples.bib}
%
\begin{document}
To \textbf{cite or not to cite?
\autocite{cicero}
\printbibliography
\end{document}
```
When this file is compiled with `latexmk -norc test` (`norc` option just to ensure no special configuration file is used), one gets the expected error:

```
Runaway argument?
{cite or not to cite? \autocite {cicero} \printbibliography \end {doc\ETC.
! File ended while scanning use of \textbf .
<inserted text> 
                \par 
<*> test.tex
```

But, if this file is corrected (`\textbf{cite` → `\textbf{cite}`), rerunning `latexmk -norc test` leads to the unexpected `biber`'s error:

```
[...]
INFO - Reading 'test.bcf'
:2241: parser error : Premature end of data in tag section line 2240

^
:2241: parser error : Premature end of data in tag controlfile line 2

^
Latexmk: Found biber source file(s) [test.bcf]
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  biber test: Command for 'biber test' gave return code 256
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
```

Would it be possible to get rid of this error?

**Comment by plk** [comment]
I think the problem is in what latexmk is doing in such circumstances. What's happening is that the broken \cite interrupts the writing of the .bcf which is therefore malformed. Manually, you would correct the \cite command a re-run latex which would create the .bcf correctly. I am not sure why latexmk doesn't re-run latex after the correction?

**Comment by dbitouze** [comment]
>I think the problem is in what latexmk is doing in such circumstances.

OK, I'll report to its author.

>Manually, you would correct the \cite

`\textbf`?

> command a re-run latex which would create the .bcf correctly.

Indeed, I should have emphasized that it is enough to rerun `latexmk`.

**Comment by plk** [comment]
Right, sorry, I meant the `\textbf`. I don't know what latexmk is doing here but another latex run after correction is mandatory to re-write the .bcf.

**Comment by dbitouze** [comment]
>Right, sorry, I meant the \textbf. 

You're welcome.

> I don't know what latexmk is doing here but another latex run after correction is mandatory to re-write the .bcf.

OK. Reported to John  Collins.

**Comment by jccollins** [comment]
The odd behavior of latexmk is because it notices that both an input file to pdflatex and an input file (.bcf) to biber have changed since the previous run of the program.  It therefore considers that both of the pdflatex and biber runs are out-of-date.  There is a circular dependence between pdflatex and biber, so under normal circumstances (i.e., in the absence of an error) there is a choice of which to run first, and it is usually best to run biber first, which latexmk is therefore programmed to do.

But in the presence of an error, this behavior can be wrong.  I need to modify latexmk to give preference to running pdflatex when the previous run ended in an error and there is a changed in a user input file (as opposed to a program-generated file).  This will be a somewhat non-trivial change.

However, this won't solve all problems of this kind.  If the incorrect latex code comes from the .bib file that biber reads, then the error can go viral.  Here is an error-causing .bib file:

    @article{Me2007,
          author = "Me",
          journal = "NEW journal",
          year = "2007",
          volume = "100",
         pages = "999",
         note = "(See also original version \textbf\cite{Me1807}.)"
    }
    @article{Me1807,
          author = "Me",
          journal = "OLD journal",
          year = "1807",
          volume = "1",
          pages = "1"
    }

and here is an error-free .tex file to use it:

    \documentclass{article}
    \usepackage[backend=biber]{biblatex}
    \addbibresource{test.bib}
    \begin{document}
    Hello.
    \cite{Me2007}
    \printbibliography
    \end{document}

Running pdflatex, biber, and then pdflatex will result in an error and give a malformed .bcf file.  Even after the error in the .bib file is corrected, no amount of running of biber and pdflatex will resolve the problem.  Running pdflatex will result in an error because of an syntax error in the .bbl file, and this gives a malformed .bcf file.  Running biber will fail because of the malformed .bcf file and will therefore fail to generate a correct .bbl file from the corrected .bib file.  The only solution is to delete both the .bcf and .bbl files and then start from a clean slate.  These kinds of error are hard to diagnose by a user, since the error message does not refer to the underlying cause of the error, but to a downstream consequence.

I think this indicates that improved error handling in biblatex and/or biber is needed.  Ideally, files generated by software (e.g., biblatex) should be well-formed.  But given the difficulties of handling error conditions inside latex packages, this may be too hard.  So I would suggest a change in biber's behavior.  When it detects a malformed .bcf file, it should generate a blank .bbl file (or one with a suitable warning) instead of simply stopping and leaving the previous .bbl file unchanged.  (Actually, latexmk could be arranged to do this.)

**Comment by plk** [comment]
That's a good idea, yes. I will see about detecting malformed .bcf files. What would be best from you point of view? I could have biber remove the .bcf completely? Would that work?

**Comment by jccollins** [comment]
I suggest that appropriate action when a malformed .bcf is detected is to 

1.  Produce an empty .bbl file (or simply delete the .bbl file).

2.  Put an appropriate error message in the .bbl file about a malformed .bcf file.

3.  Give a non-zero exit code.

I suspect there may be some ramifications of this change that need to be thought through first.  But the above seems to me to be appropriate at present.  I'll think some more about the ramifications.

The .bcf file should NOT be deleted, I think, to avoid obscuring the nature of the error.  (That was merely an action a user may currently need to take to recover from errors, given current behavior of latexmk and biber.)

**Comment by plk** [comment]
For number 2 - do you mean the .blg log file? It's probably better to delete the .bbl as per 1. Biber will write an error to screen/log anyway:

```
ERROR - test2.bcf is malformed, last biblatex run probably failed
```
 
and will exit with code 2.

**Comment by jccollins** [comment]
Yes indeed, in item 2, I of course meant the blg file.

What I meant in my comment was to list the complete set of actions needed.  Items 2 and 3 are done in the present.  It's just the deletion of the .bbl file that needs to be added.

**Comment by plk** [comment]
Ok, this is done in the dev biber 2.3/biblatex 3.2 on Sourceforge - can you try it?

**Comment by dbitouze** [comment]
Well... currently, that's worse! :)

# Previous behavior (dev `biber 2.2` and `biblatex 3.1`, `latexmk 4.43a`)
1. okay `.tex` file (`\textbf{cite}`) + `latexmk`: okay,
- buggy `.tex` file (`\textbf{cite}` → `\textbf{cite`) + `latexmk`: LaTeX error,
- okay `.tex` file (`\textbf{cite` → `\textbf{cite}`) + `latexmk`: biber error,
- okay `.tex` file (`\textbf{cite}`) + `latexmk`: okay.

# Current behavior (dev `biber 2.3` and `biblatex 3.2`, `latexmk 4.43a`)
1. okay `.tex` file (`\textbf{cite}`) + `latexmk`: okay,
- buggy `.tex` file (`\textbf{cite}` → `\textbf{cite`) + `latexmk`: LaTeX error,
- okay `.tex` file (`\textbf{cite` → `\textbf{cite}`) + `latexmk`: biber error,
- okay `.tex` file (`\textbf{cite}`) + all the subsequent `latexmk`s: (same) biber error,
- `.bcf` deletion + okay `.tex` file (`\textbf{cite}`) + `latexmk` : okay.

**Comment by plk** [comment]
Ah sorry, I believe that this needs latexmk changes too @jccollins?

**Comment by jccollins** [comment]
I'll investigate.  It may need substantial changes to the latexmk 
error-handling logic.

John

**Comment by jdleesmiller** [comment]
It may be worth noting that reverting from latexmk 4.43a to latexmk 4.39 appears to have resolved this issue for me. (I haven't tried the intervening versions.)

**Comment by nivaca** [comment]
I have this problem in latexmk 4.43a.

**Comment by jccollins** [comment]
At long last, this is fixed in latexmk 4.44.

**Comment by 153957** [comment]
Does latexmk 4.44 fix it on its own, or are the latest version of biber and biblatex also required?

**Comment by dbitouze** [comment]
>At long last, this is fixed in latexmk 4.44.

Thanks!

**Comment by plk** [comment]
biber 2.3+ is required.

**Comment by jccollins** [comment]
On 02/26/2016 04:28 AM, plk wrote:
> Closed #348 <https://github.com/plk/biblatex/issues/348>.

You should note that the solution in the combination of latexmk 4.44 and 
biber 2.3 is still not perfect.  I'll have to think some more, but I 
don't have the time right now.

The situation where there are still a problem is when there is initially 
an LaTeX error in the bib file.  After a run of biber and then 
(pdf)latex, there is a downstream error in the (pdf)latex which then 
causes a malformed bcf file.  Now suppose the user corrects the error in 
the bib file.  It turns out that latexmk won't always automatically do 
the correct set of runs that are needed.  (Why this happens and why the 
problem is so hard to solve to cover all situations would take too long 
to explain here.)

At the moment the work around in this particular situation is for the 
user to make a further small change in the .tex file, e.g., adding a 
space.  This provokes latexmk into doing things right.

John
